services/wallet/transfer/commands_sequential.go

```diff
@@ -90,152 +57,117 @@ func (c *findBlocksCommand) Run(parent context.Context) (err error) {
+	err := c.blockRangeDAO.upsertRange(c.chainClient.ChainID, c.account, blockRange)
 	if err != nil {
-		log.Info("findBlocksCommand checkRange fastIndex", "err", err)
 		c.error = err
-		// return err // In case c.noLimit is true, hystrix "max concurrency" may be reached and we will not be able to index ETH transfers
-		return nil, nil
+		log.Error("findBlocksCommand upsertRange", "error", err)
+		return err
 	}
-	log.Info("findBlocksCommand checkRange", "startBlock", startBlock, "newFromBlock", newFromBlock.Number, "toBlockNumber", to, "noLimit", c.noLimit)
 
-	// TODO There should be transfers when either when we have found headers
-	// or when newFromBlock is different from fromBlock, but if I check for
-	// ERC20 transfers only when there are ETH transfers, I will miss ERC20 transfers
+	return nil
+}
+
+func (c *findBlocksCommand) checkRange(parent context.Context, from *big.Int, to *big.Int) (
+	foundHeaders []*DBHeader, err error) {
 
-	// if len(ethHeaders) > 0 || newFromBlock.Number.Cmp(fromBlock.Number) != 0 { // there is transaction history for this account
+	fromBlock := &Block{Number: from}
 
-	erc20Headers, err := c.fastIndexErc20(parent, newFromBlock.Number, to)
+	newFromBlock, ethHeaders, startBlock, err := c.fastIndex(parent, c.balanceCache, fromBlock, to)
 	if err != nil {
-		log.Info("findBlocksCommand checkRange fastIndexErc20", "err", err)
+		log.Error("findBlocksCommand checkRange fastIndex", "err", err)
 		c.error = err
-		// return err
+		// return err // In case c.noLimit is true, hystrix "max concurrency" may be reached and we will not be able to index ETH transfers
 		return nil, nil
 	}
+	log.Debug("findBlocksCommand checkRange", "startBlock", startBlock, "newFromBlock", newFromBlock.Number, "toBlockNumber", to, "noLimit", c.noLimit)
 
-	allHeaders := append(ethHeaders, erc20Headers...)
-
-	if len(allHeaders) > 0 {
-		uniqHeadersByHash := map[common.Hash]*DBHeader{}
-		for _, header := range allHeaders {
-			uniqHeader, ok := uniqHeadersByHash[header.Hash]
-			if ok {
-				if len(header.Erc20Transfers) > 0 {
-					uniqHeader.Erc20Transfers = append(uniqHeader.Erc20Transfers, header.Erc20Transfers...)
-				}
-				uniqHeadersByHash[header.Hash] = uniqHeader
-			} else {
-				uniqHeadersByHash[header.Hash] = header
-			}
-		}
-
-		uniqHeaders := []*DBHeader{}
-		for _, header := range uniqHeadersByHash {
-			uniqHeaders = append(uniqHeaders, header)
-		}
-
-		foundHeaders = uniqHeaders
-
-		log.Info("saving headers", "len", len(uniqHeaders), "lastBlockNumber", to,
-			"balance", c.balanceCache.ReadCachedBalance(c.account, to),
-			"nonce", c.balanceCache.ReadCachedNonce(c.account, to))
-
-		err = c.db.SaveBlocks(c.chainClient.ChainID, c.account, uniqHeaders)
+	// There should be transfers when either when we have found headers
+	// or newFromBlock is different from fromBlock
+	if len(ethHeaders) > 0 || newFromBlock.Number.Cmp(fromBlock.Number) != 0 {
+		erc20Headers, err := c.fastIndexErc20(parent, newFromBlock.Number, to)
```

Comment:
Hmm I think I remember there being a good reason for calling `fastIndexErc20` regardless of the result of `fastIndex` like we currently do (that is, there was a chance of missing some blocks with transactions under some specific condition). Did you check with Roman?

Reply:
> I remember there being a good reason for calling fastIndexErc20

The reason is that we can't detect some ERC20 transfers by checking balances/nonces because they do not change those. Specifically incoming ERC20 but also some more exotic cases.

Reply:
@IvanBelyakoff if I receive 1000 USDT (for instance) this transaction will not be detected by checking balance/nonce (because it doesn't affect ETH balance and is not sent by me). **There is no good reason to not include this transaction.** Please don't do this.